[EventSource] Block EventCommandExecuted callback during initialization

mdh1418 opened the pull request:
During managed startup, the managed EventSource [`RuntimeEventSource` is initialized](https://github.com/dotnet/runtime/blob/fccbc77d2f8c32462a35488751657c8339c88508/src/coreclr/System.Private.CoreLib/src/System/StartupHookProvider.CoreCLR.cs#L19), kicking off EventSource initialization where all deferred commands `m_deferredCommands` for that EventSource instance are processed [one at a time](https://github.com/dotnet/runtime/blob/fccbc77d2f8c32462a35488751657c8339c88508/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1660) at the end of `EventSource.Initialize`.

However, should a derived EventSource set its callback `EventCommandExecuted` within its `OnEventCommand`, then because of [the decision to invoke the callback on all deferred commands](https://github.com/dotnet/runtime/blob/4fbd498cc25ca9ca2ca6cac4764866bf08d1988f/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L503-L510), the callback will be invoked before all deferred commands could be initially processed in `EventSource.Initialize`.

CounterGroup registers its callback during construction, so should multiple events be sent to `RuntimeEventSource` during initialization (e.g. `m_deferredCommands = [ Update, Update, Update, ... ]`), then because `RuntimeEventSource` [initializes multiple event counters](https://github.com/dotnet/runtime/blob/fccbc77d2f8c32462a35488751657c8339c88508/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs#L100-L131) once the deferred command is processed at the end of `EventSource.Initialize()` ((i.e. `m_deferredCommands = [ Enable, Update, Update, ... ]`), then the callback is invoked on all deferred commands before they are converted from Update.
This leads to an assertion error as demonstrated by
```
EventSource[2000003047]Invoke deferred command on this value:System.Diagnostics.Tracing.EventCommandEventArgs(-1391687009 Command = Enable)
EventSource[2000003047]Invoke deferred command on this value:System.Diagnostics.Tracing.EventCommandEventArgs(663802102 Command = Update)
```
 in https://github.com/dotnet/runtime/issues/96324#issue-2056808808.
 
This PR aims to avoid double-invocations of the same callback `m_eventCommandExecuted` during `EventSource.Initialize` should the more-derived EventSource class set-up its `EventCommandExecuted` callback within its `OnEventCommand`.

To do so, we defer invoking the `EventCommandExecuted` callback on deferred commands until **after** all deferred commands are first-processed during initialization.

Fixes #94964 
Fixes #96324
Fixes #99497
Fixes #92519

mdh1418 committed:
[EventSource] Relax assert on CounterGroup callback

tommcdon reviewed (approved):
LGTM!

noahfalk commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs:
```diff
@@ -46,8 +46,9 @@ private void RegisterCommandCallback()
-            // Should only be enable or disable
-            Debug.Assert(e.Command == EventCommand.Enable || e.Command == EventCommand.Disable);
+            // Should only be enable or disable, but during EventSource initialization, update may be seen as this callback
+            // is invoked on all m_deferredCommands while processing them synchronously through EventSource.DoCommand.
```
This comment seems technically correct, but also misleading :) If I understand the situation correctly, the callstack where we see an Update event looks like this:

```
CounterGroup.OnEventSourceCommand
EventSource.add_EventCommandExecuted
CounterGroup.RegisterCommandCallback
CounterGroup.ctor
CounterGroup.GetCounterGroup
DiagnosticCounter.Publish
PollingCounter.ctor
EventSource.OnEventCommand
EventSource.DoCommand
EventSource.Initialize
```

It is technically true that the event was received while executing EventSource.DoCommand(), but I bet most people reading the comment would assume the call is coming from [here](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,2704). Instead it is coming via this very round about call sequence.

If we leave this as-is I'd recommend clarifying in the comment about the subtle call sequence, but I think it would be better to fix the logic in EventSource that handles defered command notification. Looking at the code I think there are several potential issues:
1. The callback from EventSource.add_EventCommandExecuted can deliver different command args than the callback from DoCommand(). DoCommand() modifies Update events when it runs but add_EventCommandExecuted might dispatch them before that happens.
2. The callback from EventSource.add_EventCommandExecuted can also cause defered commands to generate two callback invocations because DoCommand() will duplicate the callback later. Not necessarily a bug, but unexpected.
3. Setting m_completelyInitialized=true [here](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,1642) means that command which arrives before Initialize() enters [this lock](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,1651) would cause all deferred commands to be [deleted](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs,2539) before we gave any notification.

I'm guessing the weird logic with m_completelyInitialized and waiting to receive another command before getting rid of the deferred commands is to let people write code like this and have their callback observe the deferred commands:
```C#
MyEventSource es = new MyEventSource();
es.EventCommandExecuted += m_callback;
```

That code already isn't guaranteed to work in the presence of a concurrently received event, but most scenarios probably don't have such an event so it turns out OK. 

I think most of weird behavior (including this bug) could be resolved by moving m_completelyInitialized = true to occur after we've dispatched the deferred commands:
```C#
lock (EventListener.EventListenersLock)
{
    EventCommandEventArgs? deferredCommands = m_deferredCommands;
    while (deferredCommands != null)
    {
        DoCommand(deferredCommands);
        deferredCommands = deferredCommands.nextCommand;
    }
    m_completelyInitialized = true;
}
```

Then we can make callback inside EventCommandExecuted conditional so that it only activates after m_completelyInitialized==true

```C#
                // If we have an EventHandler<EventCommandEventArgs> attached to the EventSource before the first command arrives
                // It should get a chance to handle the deferred commands.
                if(m_completelyInitialized)
                {
                    EventCommandEventArgs? deferredCommands = m_deferredCommands;
                    while (deferredCommands != null)
                    {
                        value(this, deferredCommands);
                        deferredCommands = deferredCommands.nextCommand;
                    }
                }
```

mdh1418 replied:
Those are very good points! I didn't even realize that there was `this.m_eventCommandExecuted?.Invoke(this, commandArgs);` since I was so focused on tracing the Assert error 😅.

I'm just realizing that EventCommandExecuted allows users to observe deferred commands in case EventSource fails to initialize. If we continue to associate `m_completelyInitialized` with initialization completed without exception (given its [comment](https://github.com/dotnet/runtime/blob/dff1d8467845fd93c517f89dc81598e5fd17c270/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L3822)) by
```C#
// Initialize post try/catch
lock (EventListener.EventListenersLock)
{
    EventCommandEventArgs? deferredCommands = m_deferredCommands;
    while (deferredCommands != null)
    {
        DoCommand(deferredCommands);
        deferredCommands = deferredCommands.nextCommand;
    }
    m_completelyInitialized = m_constructionException == null;
}
```
then guarding inside `EventCommandExecuted`
```C#
                if (m_completelyInitialized)
                {
                    EventCommandEventArgs? deferredCommands = m_deferredCommands;
                    while (deferredCommands != null)
                    {
                        value(this, deferredCommands);
                        deferredCommands = deferredCommands.nextCommand;
                    }
                }
```
will prevent that former capability. If we don't think there is any value in observing deferred commands should EventSource fail to initialize, then that is fine.
(I'm not sure what value there is in being able to see all deferred commands if the EventSource fails to initialize, unless maybe users retry EventSource initialization and requeue those deferred commands, but is EventSource flakey enough to fail initialization one time and succeed the next time?). 

However, if we want to change `m_completelyInitialized` to just mean "initialization completed" regardless of success/exception with
```C#
// Initialize post try/catch
lock (EventListener.EventListenersLock)
{
    EventCommandEventArgs? deferredCommands = m_deferredCommands;
    while (deferredCommands != null)
    {
        DoCommand(deferredCommands);
        deferredCommands = deferredCommands.nextCommand;
    }
    m_completelyInitialized = true;
}
```
then we would probably have to change SendCommand's condition to clear `m_deferredCommands` to `m_completely_initialized && m_constructionException == null` if we want to allow users to observe deferred commands if an initialization exception occurred.


I'm wondering, does `CounterGroup` call [`RegisterCommandCallback`](https://github.com/dotnet/runtime/blame/dff1d8467845fd93c517f89dc81598e5fd17c270/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs#L25) solely to set a callback for event commands sent after EventSource is completely initialized (I'm thinking this is the case), and that `EventCommandExecuted` invoking the callback for m_deferredCommands is an undesired consequence? Maybe there should be a separation between setting the `EventCommandExecuted` callback and invoking the callback on deferred commands for https://github.com/dotnet/runtime/blob/dff1d8467845fd93c517f89dc81598e5fd17c270/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L503-L504.

mdh1418 replied:
Implemented the former suggestion, so now `EventCommandExecuted` will not be able to observe deferred_commands if the EventSource fails to initialize.

noahfalk reviewed (commented):
This fix looks like it works around the issue which is OK, but I think we could fix the root cause fairly easily. Comments inline.

mdh1418 committed:
[EventSource] Avoid callback double invoke

Guard EventCommandExecuted deferred commands callbacks with
EventSource initialization to prevent CounterGroup's callback
registration from seeing unexpected commands

noahfalk commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:
```diff
@@ -2573,8 +2582,12 @@ internal void DoCommand(EventCommandEventArgs commandArgs)
+            Debug.Assert(Monitor.IsEntered(EventListener.EventListenersLock));
             // We defer commands until we are completely inited.  This allows error messages to be sent.
-            Debug.Assert(m_completelyInited);
+            // Technically, by this point initialization should be complete, but m_completelyInited
```
Similar to above, I don't think you need to make any claims about initialization being 'complete', just that m_etwProvider has been setup. That allows us to report errors if they occur.

mdh1418 replied:
Done!

noahfalk reviewed (approved):
Functionally the code change looks fine but I recommended some tweaks to the comments.

davmason reviewed (approved):
Thanks for tracking this down!

mdh1418 committed:
Simplify comments for processing deferred commands

mdh1418 commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs:
```diff
@@ -1638,8 +1641,9 @@ private unsafe void Initialize(Guid eventSourceGuid, string eventSourceName, str
                 Debug.Assert(!m_eventSourceEnabled);     // We can't be enabled until we are completely initted.
-                // We are logically completely initialized at this point.
-                m_completelyInited = true;
+                // We are logically completely initialized at this point, but set m_completelyInited after
```
Removed the comments regarding "completely initialized" to reduce confusion, and since there is a null check for `m_etwProvider` in `DoCommand`, updated the comment there
